### PR TITLE
Add device metadata folder and validation schemas

### PR DESCRIPTION
**Pull request by aparcar**
This was initially discussed over some coffee during the Hamburg 2019 meeting.

While this is a working example, tons of schema definitions are missing.

However, once this gets some traction device-tree-schema (DTS) files are validate-able via the same concept.

### CODE REVIEW

**Comment by adschm** [comment]
I'm a huge fan of this idea.

I only had a quick look for the moment. Personally, I think we could improve the folders/hierarchy/organization, I'd rather do something like
meta/schemas (for what's in scripts/schemas here, but IMO those are not scripts)
meta/devices (for what's in devices directory here)

For me, the top category is "meta", and then the difference starts on whether you have scripts, schemas, or device definitions.

File: devices/tplink_re200-v4.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+variant: v4
+bootloader: unknown
+cpu: MediaTek MT7628AN
+cpu_cores: 1
+cpu_mhz: 580
+memory_mb: 64
+storage:
+  - type: unknown
+    size_mb: 8
+device_type: WiFi Router
```
Comment (adschm) [inline devices/tplink_re200-v4.yml:12]: I'm not sure how easy it is to find reasonable and sharply distinguishable categories here.

File: devices/tplink_re200-v4.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ethernet_100m_ports: 1
+jtag: false
+serial: 57600 / 8N1
```
Comment (adschm) [inline devices/tplink_re200-v4.yml:15]: Maybe one can split this into two variables, but I'm not really sure whether that's an improvement yet.

File: devices/tplink_re200-v4.yml (hunk 1 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:12]: I used the definitions provided bye Thomas.

File: devices/tplink_re200-v4.yml (hunk 2 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:15]: We can apply the same as for storage here.

**Comment by aparcar** [comment]
> For me, the top category is "meta", and then the difference starts on whether you have scripts, schemas, or device definitions.

Sounds good to me.

File: devices/tplink_re200-v4.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+outdoor: false
```
Comment (adschm) [inline devices/tplink_re200-v4.yml:16]: As OpenWrt, I don't think this is really so much important to us; I would actually drop that parameter.

File: devices/tplink_re200-v4.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+supported_current_rel: none
+supported_since_rel: snapshot
```
Comment (adschm) [inline devices/tplink_re200-v4.yml:18]: These two would require regular updates; in this case I have a strong preference for dropping them. My idea of these files would be to not require any updates. Released devices can be tracked in other ways.

File: devices/tplink_re200-v4.yml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+target: ramips
```
Comment (adschm) [inline devices/tplink_re200-v4.yml:19]: Here, we should discuss whether we really want to put all devices into the same folder. This becomes particularly interesting when you have situations like ar71xx/ath79 transition, where the device yml file normally won't change. This would require either duplicating the file into two target subdirectories, or naming two targets in a single file.

This looks rather open to me at the moment.

File: devices/tplink_re200-v4.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+vlan: true
```
Comment (adschm) [inline devices/tplink_re200-v4.yml:20]: I'm not sure whether vlan can be limited to just true or false.

File: devices/tplink_re200-v4.yml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+wlan_24ghz: b/g
+wlan_driver: unknown
+wlan_hardware: unknown"
```
Comment (adschm) [inline devices/tplink_re200-v4.yml:23]: I wonder whether this can/should be grouped. After all, we will have modes, driver and hardware for each band. A valuable addition would be number and type of antennas btw (also per radio).

File: devices/tplink_re200-v4.yml (hunk 4 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:18]: I already deleted them from the validation as I had the same thought. Just forgot this file.

File: devices/tplink_re200-v4.yml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+vendor: TP-Link
+model: RE200
+variant: v4
+bootloader: unknown
+cpu: MediaTek MT7628AN
+cpu_cores: 1
+cpu_mhz: 580
+memory_mb: 64
```
Comment (adschm) [inline devices/tplink_re200-v4.yml:8]: Several devices support multiple cases of installed memory by autodetection (e.g. ath79/ramips).

For those, we might have e.g. 64/128 MiB variants supported with one image, i.e. appearing as one device as seen by OpenWrt. We should find a way to express this; of course we might just put "64/128", but that might break machine-readability. Can we just use the same variable name twice, or is there something like an array (I'm yaml noob)?

I'm not sure whether the latter case really helps much with machine-readability, though.

File: devices/tplink_re200-v4.yml (hunk 6 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:20]: It shouldn't what are possible values for this? 0, 16, 4096?

File: devices/tplink_re200-v4.yml (hunk 7 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:23]: I agree, just like `storage` this should be grouped

File: devices/tplink_re200-v4.yml (hunk 8 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:8]: It's possible to dynamically switch between integers and arrays of integers. That means more code for validation and e.g. rendering, but is possible.

I'm against any kind of device merging if *they are somewhat similar called different and have these random differences*, in these cases there should just be two files.

File: devices/tplink_re200-v4.yml (hunk 6 above)
Comment (adschm) [reply devices/tplink_re200-v4.yml:20]: There might be different levels of vlan awareness, and those might differ between interfaces. (I'm not an expert.)

You now mentioning a different aspect, the maximum vlan id, also tells me that the variable probably is named to generic as well. I think there is valuable information to transport here, but we discuss a bit more in which form this could be stored.

File: devices/tplink_re200-v4.yml (hunk 5 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:19]: We could use an array here as well. After all it's the same device, how we structure the target/subtargets doesn't matter. I'm against sub folders for targets which are then moved around after e.g. ar71xx deprecation.

File: devices/tplink_re200-v4.yml (hunk 3 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:16]: This was requested by Thomas, in fact there are more "optional" values which seem valuable metadata. Just as the power supply or e.g. device dimensions. I see this data source not just as a base for developers but also for tooling around it which allows users to search for specific needs.

File: devices/tplink_re200-v4.yml (hunk 5 above)
Comment (adschm) [reply devices/tplink_re200-v4.yml:19]: I'm not sure yet whether not having target subfolders is actually practical here.

However, by not having subfolders and using the target variable this could actually provide an additional benefit by providing a device name to target mapping for the user (or whatever user script is parsing this folder).

File: devices/tplink_re200-v4.yml (hunk 3 above)
Comment (adschm) [reply devices/tplink_re200-v4.yml:16]: But keep in mind that we are a router operating system/platform, and not a device feature catalogue.

I'd strongly recommend to limit this only to features somehow connected to the firmware or operation, and neither outdoor nor dimension are one of them.

File: devices/tplink_re200-v4.yml (hunk 6 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:20]: Thomas provided me with the list of Wiki values which are : Yes, No, 16, 4096. I just went with it. I'm happy to rename and convert it to a object or whatever complexity we want here.

File: devices/tplink_re200-v4.yml (hunk 5 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:19]: Maybe it's better to use the full target like `ath79/generic` instead of just the *target*, I think some devices are available in multiple subtargets(?).

File: devices/tplink_re200-v4.yml (hunk 3 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml:16]: True, I don't mean to add subjective information like `very_very_fast: False` but store static facts that are valuable to both developers and users. I'd see device specifications as documentation and therefore wouldn't be to stingy on adding it, as long as it's consistent and static.

Adding `outdoor` as an optional parameter, or actually better [ip_code](https://en.wikipedia.org/wiki/IP_Code), doesn't seem to harmful for me.

File: devices/tplink_re200-v4.yml (hunk 4 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml]: Removed.

File: devices/tplink_re200-v4.yml (hunk 6 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml]: I updated it to True, False, 16 and 4096 for now.

File: devices/tplink_re200-v4.yml (hunk 7 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml]: Updated, now it's an object with freq_ghz, driver and hardware.

**Comment by aparcar** [comment]
I did a bunch of updates, e.g. adding FriendlyARM NanoPi R1 as it has more data which should be validated: USB, keys/buttons, optional memory and storage slots. Over the next days I'll try to add descriptions everywhere so people easily understand whats happening.

File: devices/tplink_re200-v4.yml (hunk 5 above)
Comment (adschm) [reply devices/tplink_re200-v4.yml]: I also think "full" information (target and subtarget) should be provided here.

If devices are in multiple subtargets, we could either only show the target, or, (preferably) just list both, e.g. bcm63xx/generic and bcm63xx/smp like we would list ath79/generic and ar71xx/generic as well.

File: devices/tplink_re200-v4.yml (hunk 3 above)
Comment (adschm) [reply devices/tplink_re200-v4.yml]: Having an optional parameter might be a suitable compromise here. I just hesitate to make this too open, as the main openwrt repo should not become a Wiki database eventually. With optional parameters, we quickly might end up with a lot of commits just changing meta data, e.g. stuff like "add dimensions for x", "add outdoor for Y", "add case color for Z".

If we open this for optional parameters, people will want to add/move more information here, and this might get ouf of hands or by just annoying quickly. So, I'm not against adding this additional information, I mostly fear the resulting problems this will probably create.

File: devices/tplink_re200-v4.yml (hunk 2 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml]: I did not find any meaningful variation in 8N1 settings so I just kept it as a enum of strings.

File: devices/tplink_re200-v4.yml (hunk 3 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml]: Let's make dimension, power supply and ip_code non optional then?

File: devices/tplink_re200-v4.yml (hunk 5 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml]: It's like that now. Full target, an array

File: devices/tplink_re200-v4.yml (hunk 3 above)
Comment (adschm) [reply devices/tplink_re200-v4.yml]: What's ip_code and power supply?

Comment (aparcar) [reply devices/tplink_re200-v4.yml]: ip_code: https://en.wikipedia.org/wiki/IP_Code
power_supply: https://en.wikipedia.org/wiki/Power_supply

I understand that these information are not directly of concern when developing software, however so isn't the memory size when device-tree is set to auto-detection.

Having metadata maintained (additionally) by developers should improve the collaboration between wiki and developers, which are at this point still fairly separated.

Comment (adschm) [reply devices/tplink_re200-v4.yml]: > power_supply: https://en.wikipedia.org/wiki/Power_supply

Honestly?

Comment (aparcar) [reply devices/tplink_re200-v4.yml]: No this was clearly a joke, but you asked me, didn't you?

Comment (aparcar) [reply devices/tplink_re200-v4.yml]: The example section of `device.yaml` shows you how it's used:
https://github.com/openwrt/openwrt/pull/3534/files#diff-f7ab3dd4354a2c48efa41a6cb66edd8cb3469921a9233e06ae20c9984ecd0632R37

File: meta/devices/friendlyarm_nanopi-r1.yaml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  ampere: 2
+leds:
+  - color: red
+    label: status
+  - color: green
+    label: wan
+  - color: green
+    label: lan
+keys:
+  - reset
```
Comment (adschm) [inline meta/devices/friendlyarm_nanopi-r1.yaml:33]: Keys should have "function" and "mode", as there are keys and switches (linux_code = EV_SW).

File: meta/devices/friendlyarm_nanopi-r1.yaml [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  - speed_mbit: 100
+    ports: 1
+    label: lan
+power_supply:
+  type: DC
+  volt: 5
+  ampere: 2
+leds:
+  - color: red
+    label: status
```
Comment (adschm) [inline meta/devices/friendlyarm_nanopi-r1.yaml:27]: The precise term is "function" instead of "label", at least this is what they use for the new led concept upstream.

File: meta/devices/friendlyarm_nanopi-r1.yaml [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  - type: A
+    ports: 2
+ethernet:
+  - speed_mbit: 100
+    ports: 1
+    label: lan
+power_supply:
+  type: DC
+  volt: 5
+  ampere: 2
```
Comment (adschm) [inline meta/devices/friendlyarm_nanopi-r1.yaml:24]: Should that be the minimum requirement of the router or the value printed on the PSU adapter (i.e. more than necessary?)

Comment (adschm) [reply meta/devices/friendlyarm_nanopi-r1.yaml:24]: Context: E.g. outdoor devices with PoE passthrough have PSUs that provide enough power for two devices.

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 9 above)
Comment (aparcar) [reply meta/devices/friendlyarm_nanopi-r1.yaml:33]:
```suggestion
  - function: reset
    mode: switch
```

Like this?

File: meta/devices/friendlyarm_nanopi-r1.yaml [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  mhz: 1200
+memory_mb:
+  - 512
+  - 1024
+storage:
+  - type: microSD
+device_type: WiFi Router
+usb:
+  - type: A
+    ports: 2
```
Comment (adschm) [inline meta/devices/friendlyarm_nanopi-r1.yaml:16]: I'm not sure, but maybe num_ports would be better here for clarity (same for ethernet).

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 10 above)
Comment (aparcar) [reply meta/devices/friendlyarm_nanopi-r1.yaml:27]:
```suggestion
    function: status
```

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 12 above)
Comment (aparcar) [reply meta/devices/friendlyarm_nanopi-r1.yaml:16]: currently I suffixed variable names with their type, would you be okay with `port_cnt`?

Comment (aparcar) [reply meta/devices/friendlyarm_nanopi-r1.yaml:16]: Apart from that, we validation schema can also contain titles and descriptions which can explain variables in detail

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 9 above)
Comment (adschm) [reply meta/devices/friendlyarm_nanopi-r1.yaml:33]: Yes, that's at least the idea I have at the moment. I would however only add mode when it's not just a button, I don't know whether that makes sense in yaml context.

File: meta/devices/tplink_re200-v4.yml [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+cpu:
+  model: MediaTek MT7628AN
+  cores: 1
+  mhz: 580
+memory_mb:
+  - 64
+storage:
+  - type: unknown
+    size_mb: 8
+device_type: WiFi Router
```
Comment (adschm) [inline meta/devices/tplink_re200-v4.yml:14]: Hmm, with OpenWrt every device with radio is (=can by) a WiFi Router. To me this looks like it will narrow down to the intended use by the vendor, and I'm not sure whether this is really a helpful/sharp information to store.

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 11 above)
Comment (aparcar) [reply meta/devices/friendlyarm_nanopi-r1.yaml:24]: I think value printed on PSU adapter. It's hard to tell the real minimum, not?

File: meta/devices/tplink_re200-v4.yml (hunk 13 above)
Comment (aparcar) [reply meta/devices/tplink_re200-v4.yml:14]: That's true, it could be dropped

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 10 above)
Comment (adschm) [reply meta/devices/friendlyarm_nanopi-r1.yaml:27]: Yes.

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 12 above)
Comment (adschm) [reply meta/devices/friendlyarm_nanopi-r1.yaml:16]: I don't like "cnt", I'd use either port_count or port_num then.

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 11 above)
Comment (adschm) [reply meta/devices/friendlyarm_nanopi-r1.yaml:24]: >  It's hard to tell the real minimum, not?

Yes. But for my "Context" example above the value from PSU would not be so much help. Of course, this is all nitpicking. But again, this is essentially the Wiki vs. firmware-related parameter discussion.

File: devices/tplink_re200-v4.yml (hunk 6 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml]: @adschm thoughts on this?

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 12 above)
Comment (aparcar) [reply meta/devices/friendlyarm_nanopi-r1.yaml:16]: ack

File: devices/tplink_re200-v4.yml (hunk 6 above)
Comment (tmo26) [reply devices/tplink_re200-v4.yml]: Where the VLAN datafield comes from: Back in 2015 when we created the current ToH from the previous static tables, VLAN was mentioned there, so someone thought back then that it would be good to know. I have carried it with me for the past 5 years, always asking myself if it is really needed.

There was some discussion in the forum
https://forum.openwrt.org/t/maximum-number-of-vlans/34084
but I don't see a broad interest in this datafield.

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 11 above)
Comment (tmo26) [reply meta/devices/friendlyarm_nanopi-r1.yaml:24]: > Should that be the minimum requirement of the router

How would you know that? Technical specs found on OEM websites tend to be influenced by marketing guys, resulting in some OEMs providing no info at all, or even silly "110-240 VAC" while the device has a AC/DC power adapter.
We are happy about every information we get, and we are happy about every OEM that doesn't contradict himself on his website vs. owners manual etc.

Enter what is available, i.e. whatever is printed on the device itself, on the PSU, or in the owners manual.

File: meta/devices/tplink_re200-v4.yml (hunk 13 above)
Comment (tmo26) [reply meta/devices/tplink_re200-v4.yml:14]: I strongly suggest to keep the device_type. Users searching for a Wifi router do so for a reason, and they are not interested in Cameras or wired routers only.

File: devices/tplink_re200-v4.yml (hunk 3 above)
Comment (tmo26) [reply devices/tplink_re200-v4.yml]: > Having an optional parameter might be a suitable compromise here. I just hesitate to make this too open, as the main openwrt repo should not become a Wiki database eventually.

I strongly suggest to make this as open as possible, and to make this information as usefull for the users as possible. The information "outdoor yes/no" is easy to get, and highly usefull for users who are in need of an outdoor device.

File: meta/schemas/device.yaml [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+      model: slow_cpu_2
+    cpu: Atheros AR2315
+    cpu_cores: 1
+    cpu_mhz: 180
+    detachable_antennas: 1
+    device_type: WiFi Router
+    ethernet:
+      - speed_mbit: 100
+        ports: 1
+        label: lan
```
Comment (tmo26) [inline meta/schemas/device.yaml:29]: To be taken into account: There are devices out there that have e.g. 5x Gbit + 1x 100Mbit, i.e. multiple values must be possible.

Comment (aparcar) [reply meta/schemas/device.yaml]: That should work as this is an array. You could do the following:

```
ethernet:
  - speed_mbit: 100
    port_count: 1
    function: control
  - speed_mbit: 1000
    port_count: 5
    function: lan
```

File: devices/tplink_re200-v4.yml (hunk 3 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml]: Rephrasing my goal, unite efforts of developers and wiki (aka documentation) maintainers. Maybe this should be part of `openwrt.git` but of `devices.git` and then maintained by a community (e.g. device owners) but required for adding devices to `openwrt.git`.

Workflow would be:
* generate device specific YAML file
* propose it to `devices.git` or send along a PR if added via mailing list
* commit message for `openwrt.git` contains the commit ID of the `devices.git` commit that initially adds the file

This way we can get unified device information prepared by developers and then (carefully) extended by the community.

However I see a break in the current data structures. Storing latest release in the metadata means frequent updates and redundancy to the automatically created `profiles.json` files. Also all added information should be validated via schema definitions, no random and inconsistent data should be possible.

Comment (adschm) [reply devices/tplink_re200-v4.yml]: > My understanding of this PR was that this will eventually become the successor of the current dataentries / ToH. If that is not wanted (too much info unrelated to OpenWrt, too many commits for changing metadata), then we should immediately stop here and search for a different approach.

This should be discussed by more developers eventually; there might be very different opinions on whether the OpenWrt "main" repo should contain and track documentation like this. However, we can of course sharpen a few edges beforehand. Be aware that I was a fan of that idea (in general) from the start, there might be others entirely disliking it.

(Will comment on the details later.)

File: devices/tplink_re200-v4.yml (hunk 6 above)
Comment (adschm) [reply devices/tplink_re200-v4.yml]: I think both the vlan-awareness and the available number of vlans is a valuable piece of information, and it's one that is not easy to extract from the code. It's also definitely related to OpenWrt "firmware". My struggle here is mainly how to translate it into properties in a reasonable way (i.e. having both true _and_ integers for the same field is odd IMO)

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 11 above)
Comment (adschm) [reply meta/devices/friendlyarm_nanopi-r1.yaml:25]: And that's why I wouldn't provide that info in the repo.

Of course, this folds back to the question whether we will actually have the whole TOH in master; if that's the goal eventually, I would not _delete_ useful information.

BTW: how would we deal with outdated metadata in stable releases? Assuming somebody would parse it, would we backport meta data changes??

File: meta/devices/tplink_re200-v4.yml (hunk 13 above)
Comment (adschm) [reply meta/devices/tplink_re200-v4.yml:14]: But what is a WiFi router?

Would a "wired router" and a "switch" be different here, when OpenWrt can have both of them be both?

I do see that Camera or NAS are categories one can work with, but it's quite hard for devices that "only do something with ethernet/Wifi" ...

But please tell me when I'm nitpicking too much here and everybody except me knows that "Wifi router" means a box with Wifi and ethernet ports that is no Camera or NAS.

File: meta/schemas/device.yaml (hunk 14 above)
Comment (adschm) [reply meta/schemas/device.yaml]: That was also my idea of how it should work. However, "control" is something defined by the vendor which actually is not implemented that way in OpenWrt?

File: devices/tplink_re200-v4.yml (hunk 3 above)
Comment (aparcar) [reply devices/tplink_re200-v4.yml]: > Be aware that I was a fan of that idea (in general) from the start, there might be others entirely disliking it.

Yes and I'm very glad that's the case! Once this is in better shape I'll propose it to the mailing list.

File: meta/schemas/device.yaml (hunk 14 above)
Comment (aparcar) [reply meta/schemas/device.yaml]: I just came up with control. I interpret *function* as something describe the functionality or the label attached to the ports.

File: meta/devices/tplink_re200-v4.yml (hunk 13 above)
Comment (aparcar) [reply meta/devices/tplink_re200-v4.yml:14]: I think these can be considered as *consumer* categories. You can use any router for bitcoin mining but the initially designed purpose is maybe something else. For sorting this category seems fine to me.

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 12 above)
Comment (aparcar) [reply meta/devices/friendlyarm_nanopi-r1.yaml]: It's now `port_count`.

**Comment by aparcar** [comment]
Based on discussion here, on IRC and the forum with @adschm , @ynezz and @tmo26 I suggest we move this to a newly created repository called `devices.git`. The schemas should be extended to cover all required plus optional information (e.g. dimensions, IP certification), developers would be kindly asked to add their new devices to this repository. With some scripting and community efforts we can port existing devices. Whoever is of for the task (plus me) can work on a statically rendered device overview.

**Comment by aparcar** [comment]
I will keep this PR open until we have a state that's ready to become it's own repo. Also changes are now added as commits instead of force pushing. I'll eventually squash all schemas at once.

**Comment by aparcar** [comment]
I took a Wiki dump and converted it to the proposed YAML schema, the latest three commits where required to pass all of them. Find the files [here](https://github.com/aparcar/devices). I also added a `convert` script which however requires a bit of manual work.

**Comment by aparcar** [comment]
I created a (react :skull: ) webapp in the other repository which renders a form based on the schema. The filled out form then produces a valid YAML file which can be added as commit message or added to the devices repository.

https://aparcar.github.io/devices/

image

File: meta/devices/friendlyarm_nanopi-r1.yaml (hunk 11 above)
Comment (CodeFetch) [reply meta/devices/friendlyarm_nanopi-r1.yaml:25]: What is written on the device matters, because PSUs are often bulk-ware shared among different products and not necessarily the best fit in terms of energy efficiency. Please don't use the PSU values if there is no other reference. At least in the EU vendors are required to print the needed voltage/amps on the device or in the manual for these kind. The included power supply must have a specific efficiency with the device, but it can have a higher max amp value.

**Comment by tmo26** [comment]
https://github.com/openwrt/openwrt/pull/3534#discussion_r563337279

> Please don't use the PSU values if there is no other reference.

In the absence of any other information, having at least the values from the PSU is superior that having nothing at all.

**Comment by adschm** [comment]
> sence of any other information, having at least the values from the PSU is superior that having nothing at all.

Well, mixing different kinds of information without a clear statement is worse than no information, because it invalidates the data where we have a proper source.

File: meta/schemas/device-type.yaml [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+  - Camera
+  - Emulator
+  - Modem
+  - NAS
+  - Range Extender
+  - Router
+  - Single Board Computer
+  - Travel Router - Battery powered
+  - Travel Router
+  - WiFi AP
```
Comment (pepe2k) [inline meta/schemas/device-type.yaml:16]: @aparcar @adschm 
Shouldn't we use "Wi-Fi", with the dash inside, the official technology name?

Comment (CodeFetch) [reply meta/schemas/device-type.yaml:16]: @pepe2k Why not just WLAN AP? I don't like using the trademark.

Comment (adschm) [reply meta/schemas/device-type.yaml:16]: Or just "Access Point"?

**Review by tmo26 — COMMENTED** [review]
> Please don't use the PSU values if there is no other reference.

> Well, mixing different kinds of information without a clear statement is worse than no information,

Guys, I have been maintaining the ToH dataentries for the past 5 years, always in the sense of collecting data that is **useful** for users who are searching for a device that is supported by OpenWrt and that matches their criteria.

Absence of data means, that the user is not able to check a device against his criteria, i.e. missing data is of no use for the user.
Not providing anything at all, instead of the best we know (PSU data), is of no use at all.

Just for the fun of it, I dug out my first OpenWrt device: TL-MR3020 v1
Take a look at the case: https://wikidevi.wi-cat.ru/File:TL-MR3020-02.jpg

"Power" - that's it. Nothing more. Nothing on the backside or anywhere else on the device.
The PSU says _OUTPUT 5.0V= 1.0A_
(For the following, let's assume the device is not from TP-Link and no datasheet, no website, nor any other information exists about it.)

- According to your logic, no information is better than PSU data -> user can do **nothing** with missing data
- According to my logic, PSU data is better than no data -> user can do $something (i.e. **more than nothing**) with PSU data, e.g. it allows the user to estimate power consumption or search for devices with low power consumption.

In addition to this:
- In the absence of any other data: What makes you so sure that OEM datasheets / websites do always show what you would _expect_ to be written on the device? Could as well be just the PSU data instead of the device data. In the absence of any other data, you will never know, nor will you be able to verify.
- Who of the commit reviewers and approvers really verifies the provided data?
- Users do not read instructions. If they do, they have many reasons for not adhering to them. How do you make sure that the user provided data meets your expectations (device power input data, not PSU power output data)?
- Out there in real life, data availability varies wildly, as does data quality. Not taking this into account and insisting on principles will leave you with less data, and less data is less useful for the user.
- How big are the differences between PSU output and device input power data?

tl;dr: Our goal should be to collect data that is **useful for the user**. No data -> not useful for users -> goal not reached

File: meta/schemas/device-type.yaml (hunk 15 above)
Comment (pepe2k) [reply meta/schemas/device-type.yaml:16]: @CodeFetch @adschm we can have just "Access Point" but then what about the "WiFi Router" in the line below? Maybe just "Wireless" instead of WiFi/Wi-Fi/WLAN?

Comment (adschm) [reply meta/schemas/device-type.yaml:16]: @pepe2k I don't have a specific opinion here, this was just meant as a suggestion without a personal preference.
Personally, I'm not a fan of the "device-type" at all, as it's not really well-defined (as this discussion shows); I'd remove it entirely.

So, I'd leave the decision to those who actually "use" this parameter  ...

Comment (CodeFetch) [reply meta/schemas/device-type.yaml:16]: Router (without WLAN), Access Point (Router with WLAN), Ceiling AP, Outdoor AP (linearly polarized antennas), Outdoor CPE (cross-polarized antennas)... Anything else out there? 

Another "important" information for the user is if the router is "plenum-rated" and maybe if it has a metal housing and if it has a rack mounting option.

File: meta/schemas/device-type.yaml [hunk 16]
```diff
@@ -0,0 +1,19 @@
+$id: https://openwrt.org/schema/device-type.yaml
+$schema: http://json-schema.org/draft-07/schema#
+description: Type of device
+title: device-type
+type: string
+enum:
+  - Camera
+  - Emulator
```
Comment (ynezz) [inline meta/schemas/device-type.yaml:8]: Perhaps bikesheding, but what is `Emulator`? You mean `Virtual Machine` ?

File: meta/schemas/device-type.yaml [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+  - Modem
+  - NAS
```
Comment (ynezz) [inline meta/schemas/device-type.yaml:10]: You don't use `SBC` for Single Board Computer, but you do use `NAS`, maybe this should be fixed for consistency?

File: meta/schemas/device.yaml [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+    type: array
+    items:
+      $ref: file:target.yaml
+    minItems: 1
+
+  vendor:
+    $id: '#/properties/vendor'
+    $ref: file:vendor.yaml
+
+  vlan:
```
Comment (ynezz) [inline meta/schemas/device.yaml:225]: This is probably going to be more complex. This is switch property and should be modeled as that. Device can have 1-N switches, they might be of different hardware/type and thus features etc.

File: meta/schemas/device.yaml [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+    enum:
+      - BLE
+      - "4.0"
+      - "4.1"
+      - "4.2"
+      - "5.0"
+      - "5.1"
+      - "5.2"
+
+  wifi:
```
Comment (ynezz) [inline meta/schemas/device.yaml:328]: wifi, bluetooth, zigbee, lora, 802.11ah (HaLow), 802.11ad (MGWS), you-name-it are all wireless interfaces/networking standards and perhaps should be treated under one node?

File: meta/schemas/wifi.yaml [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+type: object
+
+required:
+  - freq_ghz
+  - hardware
+  - driver
+
+examples:
+  - driver: unknown
+    freq_ghz: 2.4
```
Comment (ynezz) [inline meta/schemas/wifi.yaml:13]: What about having list of supported standards like 802.11b, 802.11g etc. instead? Having information about MIMO features would be nice as well, using the popular `TxR:S` notation

* `T` Number of transmit radio chains
* `R` Number of receive radio chains
* `S` Number of spatial data streams

File: meta/schemas/wifi.yaml [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+          - unknown
+
+      hardware:
+        $id: '#/properties/wifi/items/anyOf/0/properties/hardware'
+        title: WiFi hardware
+        type: string
+
+      version:
+        $id: '#/properties/wifi/items/anyOf/0/properties/version'
+        title: WiFi Alliance Version
```
Comment (ynezz) [inline meta/schemas/wifi.yaml:59]: I would drop this marketing crap, 802.11ax is clear :P

**Comment by CodeFetch** [comment]
It would also be useful if one could describe the arrangement of the switch ports and the color of them and whether they have LEDs integrated so that we can make a visualization of the ports in LuCI.

File: meta/schemas/device-type.yaml (hunk 15 above)
Comment (pepe2k) [reply meta/schemas/device-type.yaml:16]: @adschm maybe you're right here, type of device can be in most cases easily guessed based on other, more specific information.

@CodeFetch I believe most people would say that an Access Point is a device with just single (sometimes two) Ethernet, which you connect to existing network for providing wireless access to it. I would never call a Wi-Fi router just Access Point. I'm pretty sure, if we want to be specific, we would call it... a wired router with Access Point feature :)
